roxie/ccd/ccdfile.cpp

```diff
@@ -1951,6 +2089,43 @@ class CResolvedFile : implements IResolvedFileCreator, implements ISDSSubscripti
         else
             mb.append(false);
     }
+    virtual ITranslatorSet *getTranslators(int formatCrc, IOutputMetaData *projected, IOutputMetaData *expected) const override
+    {
+        // NOTE - projected and expected and anything fetched from them such as type info may reside in dynamically loaded (and unloaded)
+        // query DLLs - this means it is not safe to include them in any sort of cache that might outlive the current query.
+        Owned<CTranslatorSet> result = new CTranslatorSet(formatCrc);
+        Owned<const IDynamicTransform> translator;
+        int prevFormatCrc = 0;
+        ForEachItemIn(idx, subFiles)
+        {
+            IOutputMetaData *actual = expected;
+            if (formatCrc)
+            {
+                const char *subname = subNames.item(idx);
+                int thisFormatCrc = formatCrcs.item(idx);
+                if (!thisFormatCrc || thisFormatCrc==formatCrc)
+                    translator.clear();
+                else
+                {
+                    actual = diskTypeInfo.item(idx);
+                    if (thisFormatCrc != prevFormatCrc)  // Check if same translation as last subfile
+                    {
+                        translator.clear();
+                        if (projected && actual)
+                        {
+                            translator.setown(createRecordTranslator(projected->queryRecordAccessor(true), actual->queryRecordAccessor(true)));
+                            // translator->describe();
+                        }
+                        if (!translator || !translator->canTranslate())
+                            throw MakeStringException(ROXIE_MISMATCH, "Untranslatable record layout mismatch detected for file %s", subname);
+                    }
+                    prevFormatCrc = thisFormatCrc;
+                }
+            }
```

Comment:
Well, really it should be creating a translator from actual to projected, regardless of what expected is, I think? Expected is only passed in as something to use for actual if the dali info did not tell us what the actual layout was.

Once it starts happening that expected and prjected differ, I think this code is still correct, but it does assume that the formatCrc from the helper refers to the projected rather than the expected format.